components/brave_ads/core/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v3.cc

```diff
@@ -0,0 +1,142 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/core/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v3.h"
+
+#include <utility>
+
+#include "base/functional/bind.h"
+#include "brave/components/brave_ads/core/internal/ads/ad_events/ad_events_database_table.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/choose/predict_ad_embeddings.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/eligible_ads/eligible_ads_features.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/eligible_ads/exclusion_rules/exclusion_rules_util.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/eligible_ads/exclusion_rules/notification_ads/notification_ad_exclusion_rules.h"
+#include "brave/components/brave_ads/core/internal/ads/serving/targeting/user_model_info.h"
+#include "brave/components/brave_ads/core/internal/ads_client_helper.h"
+#include "brave/components/brave_ads/core/internal/common/logging_util.h"
+#include "brave/components/brave_ads/core/internal/creatives/notification_ads/creative_notification_ads_database_table.h"
+#include "brave/components/brave_ads/core/internal/geographic/subdivision/subdivision_targeting.h"
+#include "brave/components/brave_ads/core/internal/resources/behavioral/anti_targeting/anti_targeting_resource.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace brave_ads::notification_ads {
+
+EligibleAdsV3::EligibleAdsV3(
```

Comment:
where v3 logic differs from v2 logic? can v2 logic be re-used?

Reply:
This will be resolved with modular ad serving pipelines. However it is important this change is backwards compatible because if we role out v3 we must be able to fall back to v2

Reply:
As I see, the code is almost the same. Maybe we can move the same code to a base class or parameterize v2?
I think we can do it without breaking v2 pipeline.

Reply:
Agreed but I would say for now we keep duplicate code as this will be removed soon when we move to modular pipelines

Reply:
As soon as this PR is merged I am working on the modular pipelines to remove all duplicated code, for now if it is ok with you I would rather keep the duplicated code. Thanks